### PR TITLE
use Object.freeze if possible instead of deepClone

### PR DESCRIPTION
**Pull request by crazy4groovy**
roboter was a strange beast... I couldn't even get it to build a dist file.

I wouldn't merge this PR if I were you, but it's a proof of concept.

### CODE REVIEW

File: lib/datasette.js [hunk 1]
```diff
@@ -5,17 +5,17 @@ const cloneDeep = require('lodash/cloneDeep'),
       defaults = require('lodash/defaults'),
       EventEmitter2 = require('eventemitter2').EventEmitter2;
 
+const frz = Object.freeze || cloneDeep;
```
Comment (crazy4groovy) [inline lib/datasette.js]: Here the main point of the PR :joy:

File: lib/datasette.test.js [hunk 2]
```diff
@@ -1,14 +1,17 @@
 'use strict';
 
-const assert = require('assertthat'),
-      cases = require('cases');
+const assert = require('assertthat');
+const cases = require('cases');
 
-const datasette = require('../../lib/datasette');
+require('uglify-js');
+require('roboter-cli');
```
Comment (crazy4groovy) [inline lib/datasette.test.js]: because roboter

**Comment by goloroden** [comment]
Hey @crazy4groovy 😊 

Thanks for the PR!

Just to make sure that I understand your intention: At the heart of your PR is using `Object.freeze` instead of lodash's `cloneDeep`, it it's available.

Can you please elaborate why this is better than staying with `cloneDeep`?

**Comment by crazy4groovy** [comment]
Hi - yes, the changes got out of control. I wandered down some rabbit holes with the testing and building tools the project is using.

I'll re-focus the PR if it's a feature you're interested in merging. Just let me know.

**Comment by goloroden** [comment]
IMHO the question is: Even if `Object.freeze` is faster, how much performance benefit do you get? And is this a module that you call 1.000 times per second? Or rarely? If the latter, then saving even a few milliseconds won't affect overall application performance at all, but the code gets more complicated.

What do you think? If you think that it's more a theoretical benefit, I'd think you can close this PR. If you think that it is a real improvement worth the additional effort in the code, feel free to adjust it.

**Comment by crazy4groovy** [comment]
It mostly depends on how much data is being deep cloned. A big object will cause a noticeable effect of app performance. (yay single thread! :))
A large number of listeners will also cause degraded performance from repeated deep cloning.

I don't think this is a theoretical problem, but each solution has its pros and cons.
Using Object.freeze won't event freeze the sub-properties of an object, so my solution isn't bulletproof either. There is a "deep freeze" npm lib you could use to overcome this issue, but performance then becomes an issue as it just recursively freezes each sub-object.

Perhaps the best option is to allow the user to pass in a config function, to either (deep) freeze the initial value, or (deep) clone it per listener call. (Default functions are just no-ops) Then the user can decide what kind of performance trade-off they are willing to accept.

Thoughts?

**Comment by goloroden** [comment]
Yes, there may be a performance problem. But OTOH the question then is whether it is the right tool for the job, or if it wouldn't make sense to store such huge objects somewhere else.

datasette is primarily meant to be used as a store for an application's configuration, not as a full-blown in-memory database. So IMHO the performance problem is not *that* important. Hence, I think we should close this PR for now, unless there is a better solution that also works with deep objects.

BTW: Using a deep freeze solves one problem, but introduces another: Since every change then requires setting up a new object, you exchange performance for memory consumption. So, either way, I guess if you need to handle huge objects, you are better suited to use a real database.

I hope that you can understand my decision to close this PR for now.
Anyway, thanks for the great discussion and the inspiration 😊

**Comment by crazy4groovy** [comment]
That's ok to close it.

P.S. Thinking a bit more, the number of listeners doesn't affect performance here as I mentioned, because you only deep clone the object once to pass into all the listeners.
However you're still open to bugs where listener 1 modifies the object and listener 2 then uses (due to object reference passing).
So deep freezing might be better than deep cloning.
Just a thought.

**Pull request closed without merging**